### PR TITLE
Buoyancy perturbation mode

### PR DESCRIPTION
**Pull request by matt-frey**

### CODE REVIEW

File: src/inversion/inversion.f90 [hunk 1]
```diff
@@ -24,49 +24,64 @@ subroutine vor2vel
...
-               svor(iz, :, :, 2) = k2l2i * (svor(iz, :, :, 2) - cs(iz, :, :))
-            enddo
-            !$omp end parallel do
+            ! Get source div(vortg) for solenoidal correction:
+            call divergence(vor, div)
 
-            ! bring back the mean x and y components of the vorticity
-            svor(:, 0, 0, 1) = ubar
-            svor(:, 0, 0, 2) = vbar
+            ! Compute vorticity correction grad(phi) where
```
Comment (daviddritschel) [inline src/inversion/inversion.f90:82]: Should be -grad(phi).  Also, I don't see where you subtract grad(phi) from the vorticity to actually correct it.

**Review by daviddritschel — COMMENTED** [review]
See my one question

File: src/inversion/inversion.f90 (hunk 1 above)
Comment (matt-frey) [reply src/inversion/inversion.f90:82]: Sorry, I am still porting over ... this got missed.

File: src/inversion/inversion.f90 [hunk 2]
```diff
@@ -243,6 +250,29 @@ subroutine buoyancy_tendency
...
+            call fftxys2p(fs, fp)
+            btend = - vel(:, :, :, 1) * fp
+            
+            call diffy(sbuoy, fs)
+            call fftxys2p(fs, fp)
+            btend = btend - vel(:, :, :, 2) * fp
+
+            call field_decompose_semi_spectral(sbuoy)
+            
+            call diffz(sbuoy, ds)
```
Comment (daviddritschel) [inline src/inversion/inversion.f90:264]: Is the result here spectral?  Are we missing a transform to physical space just after?

File: src/inversion/inversion.f90 [hunk 3]
```diff
@@ -243,6 +250,29 @@ subroutine buoyancy_tendency
...
+            call field_combine_physical(ds, fp)
+            btend = btend - vel(:, :, :, 3) * fp - bfsq * vel(:, :, :, 3)
```
Comment (daviddritschel) [inline src/inversion/inversion.f90:266]: could combine fp and vel in a bracket here

File: src/inversion/inversion.f90 [hunk 4]
```diff
@@ -440,82 +470,73 @@ end subroutine pressure
...
-            call fftxyp2s(div, ds)
+        subroutine diverge(divs, grad)
+            double precision, intent(inout)  :: divs(0:nz, 0:nx-1, 0:ny-1)
+            double precision, intent(out)    :: grad(0:nz, 0:ny-1, 0:nx-1, 3)
+            double precision                 :: us(0:nz, 0:nx-1, 0:ny-1), &
+                                                vs(0:nz, 0:nx-1, 0:ny-1), &
+                                                ws(0:nz, 0:nx-1, 0:ny-1)
 
-            ds(:, 0, 0) = zero
+            divs(:, 0, 0) = zero
```
Comment (daviddritschel) [inline src/inversion/inversion.f90:514]: I don't think we need this anymore if we are doing a 3D inversion.  In fact, when k = l = 0, we need to treat the problem differently and solve just for dphi/dz.  But to get zero boundary values, we need a condition that the integral of divs(iz, 0, 0) (discrete integral) is zero.  Seem a lot of trouble...

**Review by daviddritschel — COMMENTED** [review]
See my comments

**Review by daviddritschel — APPROVED** [review]

**Pull request merged**